guides/keras_nlp/neural_machine_translation.py

```diff
@@ -0,0 +1,442 @@
+eng_tokens_ex = eng_tokenizer.tokenize(eng_input_ex)
+print("English sentence: ", eng_input_ex)
+print("Tokens: ", eng_tokens_ex)
+print("Recovered text after detokenizing: ", eng_tokenizer.detokenize(eng_tokens_ex))
+
+print()
+
+spa_input_ex = text_pairs[0][1]
+spa_tokens_ex = spa_tokenizer.tokenize(spa_input_ex)
+print("Spanish sentence: ", spa_input_ex)
+print("Tokens: ", spa_tokens_ex)
+print("Recovered text after detokenizing: ", spa_tokenizer.detokenize(spa_tokens_ex))
+
+"""
+## Format Datasets
+
+Next, we'll format our datasets.
+
+At each training step, the model will seek to predict target words N+1 (and beyond)
+using the source sentence and the target words 0 to N.
+
+As such, the training dataset will yield a tuple `(inputs, targets)`, where:
+
+- `inputs` is a dictionary with the keys `encoder_inputs` and `decoder_inputs`.
+`encoder_inputs` is the vectorized source sentence and `decoder_inputs` is the target sentence "so far",
+that is to say, the words 0 to N used to predict word N+1 (and beyond) in the target sentence.
+- `target` is the target sentence offset by one step:
+it provides the next words in the target sentence -- what the model will try to predict.
+
+Before we tokenize the text, we will add [START] and [END] tokens to the input
+Spanish sentence.
+"""
+
+train_pairs = [
+    (text_pair[0], "[START] " + text_pair[1] + " [END]") for text_pair in train_pairs
+]
+val_pairs = [
+    (text_pair[0], "[START] " + text_pair[1] + " [END]") for text_pair in val_pairs
+]
+test_pairs = [
+    (text_pair[0], "[START] " + text_pair[1] + " [END]") for text_pair in test_pairs
+]
+
+
+def format_dataset(eng, spa):
+    eng = eng_tokenizer(eng)
+    spa = spa_tokenizer(spa)
+    return (
+        {"encoder_inputs": eng, "decoder_inputs": spa[:, :-1],},
+        spa[:, 1:],
+    )
+
+
+def make_dataset(pairs):
+    eng_texts, spa_texts = zip(*pairs)
+    eng_texts = list(eng_texts)
+    spa_texts = list(spa_texts)
+    dataset = tf.data.Dataset.from_tensor_slices((eng_texts, spa_texts))
+    dataset = dataset.batch(BATCH_SIZE)
+    dataset = dataset.map(format_dataset)
+    return dataset.shuffle(2048).prefetch(16).cache()
+
+
+train_ds = make_dataset(train_pairs)
+val_ds = make_dataset(val_pairs)
+
+"""
+Let's take a quick look at the sequence shapes
+(we have batches of 64 pairs, and all sequences are 40 steps long):
+"""
+
+for inputs, targets in train_ds.take(1):
+    print(f'inputs["encoder_inputs"].shape: {inputs["encoder_inputs"].shape}')
+    print(f'inputs["decoder_inputs"].shape: {inputs["decoder_inputs"].shape}')
+    print(f"targets.shape: {targets.shape}")
+
+
+"""
+## Building the model
+
+Now, let's move on to the exciting part - defining our model!
+We first need an Embedding layer, i.e., a vector for every token in our input sequence.
+This Embedding layer can be initialised randomly. We also need a Positional
+Embedding layer which encodes the word order in the sequence. The convention is
```

Comment:
Again, use backticks for code keywords everywhere.

Reply:
Same comment as before.